packages/@react-spectrum/overlays/src/Popover.tsx

```diff
@@ -115,54 +133,67 @@ const PopoverWrapper = forwardRef((props: PopoverWrapperProps, ref: RefObject<HT
         {!isNonModal && <DismissButton onDismiss={state.close} />}
         {children}
         {hideArrow ? null : (
-          <Arrow arrowProps={arrowProps} direction={arrowPlacement[placement]} />
+          <Arrow
+            arrowProps={arrowProps}
+            isLandscape={arrowPlacement[placement] === 'bottom'}
+            arrowRef={arrowRef}
+            primary={primary}
+            secondary={secondary}
+            borderDiagonal={borderDiagonal} />
         )}
         <DismissButton onDismiss={state.close} />
       </div>
     </div>
   );
 });
 
-let ROOT_2 = Math.sqrt(2);
+function usePopoverBorderRadius(popoverRef: RefObject<HTMLDivElement>) {
+  let [borderRadius, setBorderRadius] = useState(0);
+  useLayoutEffect(() => {
+    if (popoverRef.current) {
+      let spectrumBorderRadius = window.getComputedStyle(popoverRef.current).borderRadius;
+      if (spectrumBorderRadius !== '') {
+        setBorderRadius(parseInt(spectrumBorderRadius, 10));
+      }
+    }
+  }, [popoverRef]);
+  return borderRadius;
+}
 
-function Arrow(props) {
+function useArrowSize() {
   let [size, setSize] = useState(20);
   let [borderWidth, setBorderWidth] = useState(1);
-  let ref = useRef();
+  let arrowRef = useRef<SVGSVGElement>(null);
   // get the css value for the tip size and divide it by 2 for this arrow implementation
   useLayoutEffect(() => {
-    if (ref.current) {
-      let spectrumTipWidth = window.getComputedStyle(ref.current)
+    if (arrowRef.current) {
+      let spectrumTipWidth = window.getComputedStyle(arrowRef.current)
         .getPropertyValue('--spectrum-popover-tip-size');
       if (spectrumTipWidth !== '') {
         setSize(parseInt(spectrumTipWidth, 10) / 2);
       }
 
-      let spectrumBorderWidth = window.getComputedStyle(ref.current)
+      let spectrumBorderWidth = window.getComputedStyle(arrowRef.current)
         .getPropertyValue('--spectrum-popover-tip-borderWidth');
       if (spectrumBorderWidth !== '') {
         setBorderWidth(parseInt(spectrumBorderWidth, 10));
       }
     }
-  }, [ref]);
-
-  let landscape = props.direction === 'top' || props.direction === 'bottom';
-  let mirror = props.direction === 'left' || props.direction === 'top';
-
-  let borderDiagonal = borderWidth * ROOT_2;
-  let halfBorderDiagonal = borderDiagonal / 2;
+  }, []);
+  return {size, borderWidth, arrowRef};
+}
 
-  let secondary = 2 * size + 2 * borderDiagonal;
-  let primary = size + borderDiagonal;
+function Arrow(props: ArrowProps) {
```

Comment:
Why wasn't `props` deconstructed? I see `primary`, `secondary` and `isLandscape` each used a few times.

Reply:
just old code from main, will update